Adding tf coil with rounded corneres

mateczentye opened the pull request:
## Proposed changes

Added new rectangle toroidal field coil class with rounded corners. 
Added new test file to run tests on the class. The additional 17 tests for the class checks input parameters and compares parametric volume and surface area with analytical results for a pre-set values. The analytical calculation carried out within the test file via two functions.
Added pytest.ini to avoid warnings from custom markers, which were used to selectively run the tests locally. It can also be used in different CI setups which I think could prove to be useful.

## Types of changes

What types of changes does your code introduce to the Paramak?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [x] Documentation Update (if none of the other choices apply)
- [x] New tests

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Pep8 applied
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

Shimwell commented:
@mateczentye there are a few new things here, thanks for the PR. Would you be able to explain the new features and in particular the new pytest parts.

Thanks

mateczentye commented:
> @mateczentye there are a few new things here, thanks for the PR. Would you be able to explain the new features and in particular the new pytest parts.
> 
> Thanks

Thanks @Shimwell! I have updated the description to reflect what I have added

Shimwell commented:
Code coverage report suggests the new component class is not being covered by the new tests

Shimwell commented:
> Code coverage report suggests the new component class is not being covered by the new tests

Sorry ignore that comment, it looks like most of the new class is covered by tests and code cov is picking this up

Shimwell commented:
I was just cleaning up the doc string so that it works with read the docs and noticed we are missing the docs for this nice new component.

Could we have a few lines added to the file /docs/source/paramak.parametric_components.rst

Shimwell commented on paramak/parametric_components/toroidal_field_coil_round_corners.py:
```diff
@@ -0,0 +1,344 @@
+
+        return solid
+
+
+if __name__ == "__main__":
+
+    obj = ToroidalFieldCoilRectangleRoundCorners(
+        lower_inner_coordinates=(50, 0),
+        mid_point_coordinates=(100, 100),
+        thickness=20,
+        distance=20,
+        number_of_coils=1,
+        rotation_angle=180,
+        with_inner_leg=True,
+        test=True,
+        analyse=True
+    )
```
```suggestion
```

Shimwell replied:
I think the intended use of file is via an import like ```paramak.ToroidalFieldCoilRectangleRoundCorners()``` not running directly using __main__

Shimwell commented on paramak/parametric_components/toroidal_field_coil_round_corners.py:
```diff
@@ -0,0 +1,344 @@
+        stp_filename: Optional[str] = "ToroidalFieldCoilRectangleRoundCorners.stp",
+        stl_filename: Optional[str] = "ToroidalFieldCoilRectangleRoundCorners.stl",
+        material_tag: Optional[str] = "outter_tf_coil_mat",
+        test: Optional[bool] = False,
```
```suggestion
```

Shimwell replied:
I think we can remove the test flag as the points can be found via ```ToroidalFieldCoilRectangleRoundCorners.points```

Shimwell commented on paramak/parametric_components/toroidal_field_coil_round_corners.py:
```diff
@@ -0,0 +1,328 @@
+        file_name_stp: Defults to "ToroidalFieldCoilRectangleRoundCorners.stp"
+        file_name_stl: Defaults to "ToroidalFieldCoilRectangleRoundCorners.stl"
+        material_tag: Defaults to "outter_tf_coil_mat"
+        test: True prints to console the list of points
```
```suggestion
```

Shimwell commented on paramak/parametric_components/toroidal_field_coil_round_corners.py:
```diff
@@ -0,0 +1,328 @@
+        self.inner_leg_connection_points = [
+            inner_p1, inner_p2, inner_p3, inner_p4]
+
+        # For self testing and debugging
+        if self.test:
+            print(tri_points)
```
```suggestion
```

mateczentye replied:
I now removed the test and analyse Booleans 😄 removed their tests as well, and checked the analyse_attributes are being called in right in the test file so they were not needed.

Shimwell commented:
I think if I have understood what the ```analyse``` option is doing then we also don't need it in the class as one can access volumes and areas using the ```ToroidalFieldCoilRectangleRoundCorners.area``` ```ToroidalFieldCoilRectangleRoundCorners.areas```
```ToroidalFieldCoilRectangleRoundCorners.volume```
```ToroidalFieldCoilRectangleRoundCorners.volumes```

Shimwell commented:
@mateczentye CI runs on ever commit for an open PR. So I'm goign to got to actions and manually cancel a few of the older ones

mateczentye commented:
> @mateczentye CI runs on ever commit for an open PR. So I'm goign to got to actions and manually cancel a few of the older ones

No worries, yes sorry been trying to get things fixed 😄

Shimwell commented:
> > @mateczentye CI runs on ever commit for an open PR. So I'm goign to got to actions and manually cancel a few of the older ones
> 
> No worries, yes sorry been trying to get things fixed 😄

We should change the CI at some point so that only tagged pushes trigger it instead of every push. I managed to cancel 11 CI runners

Shimwell commented:
Looks like ```analyse``` appears in a few places is the plan to fully take this out

Shimwell commented:
@mateczentye did you notice I've added an automatic code inspections which provide some nice insights. Are you able to address a few 

https://github.com/fusion-energy/paramak/pull/27/checks?check_run_id=3084498221

mateczentye commented:
> @mateczentye did you notice I've added an automatic code inspections which provide some nice insights. Are you able to address a few
> 
> https://github.com/fusion-energy/paramak/pull/27/checks?check_run_id=3084498221

yes I think I have fixed them now, except the __init__ file which I do in a second 😄

mateczentye commented:
@Shimwell Code inspector is now passing, handy little tool! Good for getting a good practice at using conventions! 😄

Shimwell commented:
With the way type checking is done I think the parameters are only checked when originally input

This would be checked
```my_coil = ToroidalFieldCoilRectangleRoundCorners(distance=10)```

But subsequent changes would not be checked
```my_coil.distance = 20```

perhaps make use of getters and setters instead of checking in the init

Shimwell commented:
Wondering if we can just remove duplicate points from the list of points instead of doing the 10% thing

Shimwell commented:
> @Shimwell Code inspector is now passing, handy little tool! Good for getting a good practice at using conventions! 😄

Thanks for working through that, I plan to add a badge with the project score so making sure the code base passes this test is useful.

mateczentye commented:
> Wondering if we can just remove duplicate points from the list of points instead of doing the 10% thing

If I remove this and just rid of the points the inner corner would make a rectangle corner when the thickness is larger than the base length (the difference of the x-coordinates of the two given tuples for mid and low points). In general its there if there is some arbitrary large thickness is passed compared to the base length and to avoid breaking the code or making the said rectangle corner. The other thing I can do is revisit the calculation and make it a different function so there would be no need for the logical statement at all.

Shimwell commented:
> > Wondering if we can just remove duplicate points from the list of points instead of doing the 10% thing
> 
> If I remove this and just rid of the points the inner corner would make a rectangle corner when the thickness is larger than the base length (the difference of the x-coordinates of the two given tuples for mid and low points). In general its there if there is some arbitrary large thickness is passed compared to the base length and to avoid breaking the code or making the said rectangle corner. The other thing I can do is revisit the calculation and make it a different function so there would be no need for the logical statement at all.

Ah ok this sounds tricky, lets leave this part as it is. We can change it later if needed.

Shimwell commented:
> > @Shimwell Code inspector is now passing, handy little tool! Good for getting a good practice at using conventions! 😄
> 
> Thanks for working through that, I plan to add a badge with the project score so making sure the code base passes this test is useful.

Badges are being added over here https://github.com/fusion-energy/paramak/pull/28

Shimwell commented:
> With the way type checking is done I think the parameters are only checked when originally input
> 
> This would be checked
> `my_coil = ToroidalFieldCoilRectangleRoundCorners(distance=10)`
> 
> But subsequent changes would not be checked
> `my_coil.distance = 20`
> 
> perhaps make use of getters and setters instead of checking in the init

I think this is the only remaining issue is the type checking via getters and setters instead of in the init. Are you happy to make these changes then we can merge

mateczentye commented:
> > With the way type checking is done I think the parameters are only checked when originally input
> > This would be checked
> > `my_coil = ToroidalFieldCoilRectangleRoundCorners(distance=10)`
> > But subsequent changes would not be checked
> > `my_coil.distance = 20`
> > perhaps make use of getters and setters instead of checking in the init
> 
> I think this is the only remaining issue is the type checking via getters and setters instead of in the init. Are you happy to make these changes then we can merge

Yes just finishing up the documentation part and then will set the getters and setters up for the variables and make the tests for them 😄

Shimwell commented:
> > > With the way type checking is done I think the parameters are only checked when originally input
> > > This would be checked
> > > `my_coil = ToroidalFieldCoilRectangleRoundCorners(distance=10)`
> > > But subsequent changes would not be checked
> > > `my_coil.distance = 20`
> > > perhaps make use of getters and setters instead of checking in the init
> > 
> > 
> > I think this is the only remaining issue is the type checking via getters and setters instead of in the init. Are you happy to make these changes then we can merge
> 
> Yes just finishing up the documentation part and then will set the getters and setters up for the variables and make the tests for them 😄

Thanks I forgot about the docs, nice one

mateczentye commented:
![TFcoil(1)](https://user-images.githubusercontent.com/85617935/125977357-3b893496-fc6b-4f06-8a95-10164cb70d7c.png)

mateczentye commented:
![Screenshot from 2021-07-16 17-20-39](https://user-images.githubusercontent.com/85617935/125979064-c9aff900-4a07-462d-91bd-53af3bc66559.png)
![Screenshot from 2021-07-16 17-22-46](https://user-images.githubusercontent.com/85617935/125979068-7541a0ff-54ef-45ba-a465-ef62850675f4.png)

mateczentye commented:
![Screenshot from 2021-07-16 17-25-45](https://user-images.githubusercontent.com/85617935/125979248-1dbbd7be-916b-4921-86ed-a942b6cda757.png)

pulled wrong one in for earlier comment, this one has the outer edges highlighted

Shimwell commented:
Nice progress on the dos, looking at the build it appears there is some missing white space for the images and the shinx buiild is not quite working

https://paramak--27.org.readthedocs.build/en/27/paramak.parametric_components.html#toroidalfieldcoilrectangleroundcorners

Shimwell commented on docs/source/paramak.parametric_components.rst:
```diff
@@ -752,3 +752,36 @@ CapsuleVacuumVessel()
    :members:
    :show-inheritance:
+
+ToroidalFieldCoilRectangleRoundCorners()
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. cadquery::
+   :select: cadquery_object
+   :gridsize: 0
+
+   import paramak
+   my_component = paramak.ToroidalFieldCoilRectangleRoundCorners(
+      with_inner_leg=False,
+      lower_inner_coordinates=(0, 0),
+      mid_point_coordinates=(100, 100),
+      thickness=30,
+      number_of_coils=1,
+      distance=20,
+      rotation_angle=180,
+   )
+
+   cadquery_object = my_component.solid
+
+|TFCoilRoundCornersvg| |TFCoilRoundCornersvg2| |TFCoilRoundCornerstp|
+
+..|TFCoilRoundCornersvg| image: https://user-images.githubusercontent.com/85617935/125979064-c9aff900-4a07-462d-91bd-53af3bc66559.png
+  :width: 255
+..|TFCoilRoundCornersvg2| image: https://user-images.githubusercontent.com/85617935/125979248-1dbbd7be-916b-4921-86ed-a942b6cda757.png
+  :width: 255
+..|TFCoilRoundCornersvg| image: https://user-images.githubusercontent.com/85617935/125977357-3b893496-fc6b-4f06-8a95-10164cb70d7c.png
+  :width: 255
+
+.. automodule:: paramak.parametric_components.toroidal_field_coil_round_corners
+   :members:
+   :show-inheritance:
+
```
```suggestion
ToroidalFieldCoilRectangleRoundCorners()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.. cadquery::
   :select: cadquery_object
   :gridsize: 0

   import paramak
   my_component = paramak.ToroidalFieldCoilRectangleRoundCorners(
      with_inner_leg=False,
      lower_inner_coordinates=(0, 0),
      mid_point_coordinates=(100, 100),
      thickness=30,
      number_of_coils=1,
      distance=20,
      rotation_angle=180,
   )

   cadquery_object = my_component.solid

|TFCoilRoundCornersvg| |TFCoilRoundCornersvg2| |TFCoilRoundCornerstp|

.. |TFCoilRoundCornersvg| image: https://user-images.githubusercontent.com/85617935/125979064-c9aff900-4a07-462d-91bd-53af3bc66559.png
  :width: 255
.. |TFCoilRoundCornersvg2| image: https://user-images.githubusercontent.com/85617935/125979248-1dbbd7be-916b-4921-86ed-a942b6cda757.png
  :width: 255
.. |TFCoilRoundCornersvg| image: https://user-images.githubusercontent.com/85617935/125977357-3b893496-fc6b-4f06-8a95-10164cb70d7c.png
  :width: 255

.. automodule:: paramak.parametric_components.toroidal_field_coil_round_corners
   :members:
   :show-inheritance:

```

Shimwell commented on docs/source/paramak.parametric_components.rst:
```diff
@@ -752,3 +752,35 @@ CapsuleVacuumVessel()
+
+ToroidalFieldCoilRectangleRoundCorners()
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. cadquery::
```
```suggestion

.. cadquery::
```

mateczentye commented:
> Nice progress on the dos, looking at the build it appears there is some missing white space for the images and the shinx buiild is not quite working
> 
> https://paramak--27.org.readthedocs.build/en/27/paramak.parametric_components.html#toroidalfieldcoilrectangleroundcorners

@Shimwell thank you for fixing the docs, I have now added getters and setters for the attributes and wrote tests for the setters. I reshuffled the test file too so should be easier to read. Wasn't sure if we needed tests for getters, so if we need it let me know I can make them. I think I now fixed code inspector problems too but will see 😄

Shimwell commented on paramak/parametric_components/toroidal_field_coil_round_corners.py:
```diff
@@ -0,0 +1,371 @@
+            0
+        ]
+
+        ### Check if input values are what they meant to be ###
+        if not isinstance(self._lower_inner_coordinates, tuple):
+            raise TypeError("Invalid input - Coordinates must be a tuple")
+
+        if not isinstance(self._mid_point_coordinates, tuple):
+            raise TypeError("Invalid input - Coordinates must be a tuple")
+
+        if not isinstance(self._thickness, float):
+            if not isinstance(self._thickness, int):
+                raise TypeError("Invalid input - Thickness must be a number")
+
+        if not isinstance(self._distance, float):
+            if not isinstance(self._distance, int):
+                raise TypeError("Invalid input - Distance must be a number")
+
+        if (self._number_of_coils % 1) != 0:
+            raise TypeError(
+                "Invalid input - Number of Coils must be an integer number")
+
+        if len(lower_inner_coordinates) != 2 or len(
+                mid_point_coordinates) != 2:
+            raise ValueError(
+                "The input tuples are too long or too short, they must be 2 element long")
+
+        if self._lower_inner_coordinates[0] > self._mid_point_coordinates[0]:
+            raise ValueError(
+                "The middle point's x-coordinate must be larger than the lower",
+                "inner point's x-coordinate")
+
```
```suggestion
```

Shimwell replied:
Can these now be removed as the setters check thin inputs are correct?

Shimwell commented on paramak/parametric_components/toroidal_field_coil_round_corners.py:
```diff
@@ -0,0 +1,342 @@
+        this to avoid having coordinates before the previous or at the same spot as Paramak
+        cannot compute it"""
+
+        if self._thickness / self._base_length >= 1:
+            self._inner_curve_radius = self._thickness * 0.1
+            self._outter_curve_radius = self._thickness * 1.1
+            self._analyse_attributes[2] = self._inner_curve_radius
+            self._analyse_attributes[3] = self._outter_curve_radius
+        else:
+            self._outter_curve_radius = (
+                1 + (self._thickness / self._base_length)) * self._thickness
+            self._inner_curve_radius = (self._thickness**2) / self._base_length
+            self._analyse_attributes[2] = self._inner_curve_radius
+            self._analyse_attributes[3] = self._outter_curve_radius
```
This will only be run in initialization, it won't be run if the values are changed later. Perhaps this should be moved elsewhere

mateczentye replied:
I have now moved that into a new method which is being called in the constructor and the setter. I put 2 conditions back to check initialising parameters for the tuples so now all tests run

billingsley-john commented:
Looks good. Upon removing Lines 71 to 98, should lines 57 to 70 also be updated to set visible variables instead of hidden variables?

Shimwell commented on paramak/parametric_components/toroidal_field_coil_round_corners.py:
```diff
@@ -0,0 +1,369 @@
+            **kwargs
+        )
+
+        self._lower_inner_coordinates = lower_inner_coordinates[0], lower_inner_coordinates[1]
+        self._mid_point_coordinates = mid_point_coordinates[0], mid_point_coordinates[1]
```
```suggestion
        self._lower_inner_coordinates = lower_inner_coordinates
        self._mid_point_coordinates = mid_point_coordinates
```

Shimwell replied:
Is there any need to separate these two out ?

mateczentye replied:
it was my  initial way of getting around the tuple length issue if there would be any. It doesn't have to be separate anymore

Shimwell commented on paramak/parametric_components/toroidal_field_coil_round_corners.py:
```diff
@@ -0,0 +1,369 @@
+                "The middle point's x-coordinate must be larger than the lower",
+                "inner point's x-coordinate")
+
+        # Adding hidden attributes for analyse list population
+        # inner base length of the coil
+        self._base_length = self._mid_point_coordinates[0] - \
+            self._lower_inner_coordinates[0]
+        self._analyse_attributes[0] = self._base_length
+
+        # height of the coil
+        self._height = abs(
+            mid_point_coordinates[1] - lower_inner_coordinates[1]) * 2
+        self._analyse_attributes[1] = self._height
```
```suggestion
        # Adding hidden attributes for analyse list population
        # inner base length of the coil
        self._base_length = self._mid_point_coordinates[0] - \
            self._lower_inner_coordinates[0]
        self._analyse_attributes[0] = self._base_length

        # height of the coil
        self._height = abs(
            mid_point_coordinates[1] - lower_inner_coordinates[1]) * 2
        self._analyse_attributes[1] = self._height
```

These attributes might get out of date if they are set in the init and then changed later.

Shimwell replied:
perhaps they can be moved to being calculated just before they are needed

mateczentye replied:
Good point, I can move them into a method like the find_radii and just call them were needed in a similar way of that method?

Shimwell commented on paramak/parametric_components/toroidal_field_coil_round_corners.py:
```diff
@@ -0,0 +1,369 @@
+            mid_point_coordinates[1] - lower_inner_coordinates[1]) * 2
+        self._analyse_attributes[1] = self._height
+
+        self._find_radii()
```
```suggestion
        self._find_radii()
```

Is there any need to call find_radii here, seeing as it is called in find_points

mateczentye replied:
I tried it only being inside the find_points, but then with object initialisation it appears not to calculate it if I do this in a test suit:
print(object.analyse_attributes)
object.show()

I'd have to call the create_solid in to have the calculation done before printing if its only in the find_points and wasn't sure if that is an effective way of getting them.

Shimwell replied:
Do we need ```analyse_attributes``` at all. It appears to just be a list of other attributes.

mateczentye replied:
it is used in the analytical volume and surface calculations in the testing file

Shimwell replied:
The getter for ```analyse_attributes``` can be set to call ```find_points``` if that helps

mateczentye replied:
oh yes very good point 😄 I make that change now

Shimwell commented on paramak/parametric_components/toroidal_field_coil_round_corners.py:
```diff
@@ -0,0 +1,369 @@
+            0
+        ]
+
+        if len(lower_inner_coordinates) != 2 or len(
+                mid_point_coordinates) != 2:
+            raise ValueError(
+                "The input tuples are too long or too short, they must be 2 element long")
```
This looks like it duplicates a check that is done in the setter, perhaps this can be removed
```suggestion
```

mateczentye replied:
I removed it, but then failed two of the tests when initialising an object so I put it back.

Shimwell commented on paramak/parametric_components/toroidal_field_coil_round_corners.py:
```diff
@@ -0,0 +1,369 @@
+        # this to avoid having coordinates before the previous or at the same spot as Paramak
+        # cannot compute it
+
+        if self._thickness / self._base_length >= 1:
+            self._inner_curve_radius = self._thickness * 0.1
+            self._outter_curve_radius = self._thickness * 1.1
+            self._analyse_attributes[2] = self._inner_curve_radius
+            self._analyse_attributes[3] = self._outter_curve_radius
+        else:
+            self._outter_curve_radius = (
+                1 + (self._thickness / self._base_length)) * self._thickness
+            self._inner_curve_radius = (self._thickness**2) / self._base_length
+            self._analyse_attributes[2] = self._inner_curve_radius
+            self._analyse_attributes[3] = self._outter_curve_radius
```
```suggestion
        if self._thickness / self._base_length >= 1:
            self._inner_curve_radius = self._thickness * 0.1
            self._outter_curve_radius = self._thickness * 1.1
        else:
            self._outter_curve_radius = (
                1 + (self._thickness / self._base_length)) * self._thickness
            self._inner_curve_radius = (self._thickness**2) / self._base_length

        self._analyse_attributes[2] = self._inner_curve_radius
        self._analyse_attributes[3] = self._outter_curve_radius
```

Shimwell replied:
Looks like there was some duplication here

Shimwell commented on tests/test_parametric_components/test_toroidalfieldcoilrectangleroundcorner.py:
```diff
@@ -0,0 +1,392 @@
+    with pytest.raises(TypeError):
+        obj.distance = "string"
+
+# Number of Coils
+
```
```suggestion
```

Shimwell replied:
removed random comment. We do have doc strings for explaining the tests but they go just under the def section with triple quotes.

Shimwell replied:
I see there are a few of these comments in the test script. Any chance they could be moved to doc strings

Shimwell replied:
Ah ignore this, I can perhaps just change the spacing a bit then I think we can merge. Lots of changes already so I don't want to be fussy :-)

mateczentye commented:
@Shimwell thank you for the merge! 😃
